kolibri/core/discovery/utils/network/search.py

```diff
@@ -0,0 +1,175 @@
+    )
+    return name.replace(SERVICE_TYPE, "").strip(".")
+
+
+def _is_port_open(host, port, timeout=1):
+    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
+        sock.settimeout(timeout)
+        return sock.connect_ex((host, port)) == 0
+
+
+class KolibriZeroconfService(object):
+
+    info = None
+
+    def __init__(self, id, port=8080, data={}):
+        self.id = id
+        self.port = port
+        self.data = {key: json.dumps(val) for (key, val) in data.items()}
+        atexit.register(self.cleanup)
+
+    def register(self):
+
+        if not ZEROCONF_STATE["zeroconf"]:
+            initialize_zeroconf_listener()
+
+        assert self.info is None, "Service is already registered!"
+
+        i = 1
+        id = self.id
+
+        while not self.info:
+
+            # attempt to create an mDNS service and register it on the network
+            try:
+                info = ServiceInfo(
+                    SERVICE_TYPE,
+                    name=".".join([id, SERVICE_TYPE]),
+                    server=".".join([id, LOCAL_DOMAIN, ""]),
+                    address=USE_IP_OF_OUTGOING_INTERFACE,
+                    port=self.port,
+                    properties=self.data,
+                )
+
+                ZEROCONF_STATE["zeroconf"].register_service(info, ttl=60)
+
+                self.info = info
+
+            except NonUniqueNameException:
+                # if there's a name conflict, append incrementing integer until no conflict
+                i += 1
+                id = "%s-%d" % (self.id, i)
+
+            if i > 100:
+                raise NonUniqueNameException()
+
+        self.id = id
+
+        return self
+
+    def unregister(self):
+
+        assert self.info is not None, "Service is not registered!"
+
+        ZEROCONF_STATE["zeroconf"].unregister_service(self.info)
+
+        self.info = None
+
+    def cleanup(self, *args, **kwargs):
+
+        if self.info and ZEROCONF_STATE["zeroconf"]:
+            self.unregister()
+
+
+class KolibriZeroconfListener(object):
+
+    instances = {}
+
+    def add_service(self, zeroconf, type, name):
+        info = zeroconf.get_service_info(type, name)
+        id = _id_from_name(name)
+        ip = socket.inet_ntoa(info.address)
+        self.instances[id] = {
+            "id": id,
+            "ip": ip,
+            "local": ip in get_all_addresses(),
+            "port": info.port,
+            "host": info.server.strip("."),
+            "data": {key: json.loads(val) for (key, val) in info.properties.items()},
+            "base_url": "http://{ip}:{port}/".format(ip=ip, port=info.port),
+        }
+        logger.info(
+            "Kolibri instance '%s' joined zeroconf network; service info: %s\n"
+            % (id, self.instances[id])
+        )
+
+    def remove_service(self, zeroconf, type, name):
+        id = _id_from_name(name)
+        logger.info("\nKolibri instance '%s' has left the zeroconf network.\n" % (id,))
```

Comment:
The newline at the beginning seems wrong?

Reply:
Oh and the assigned to a builtin once again. @indirectlylit I think that removing Flake 8 was a mistake, since it checks many things that Black doesn't check. Assignments, code complexity, unused variables.

![image](https://user-images.githubusercontent.com/374612/59915349-b64cea80-941c-11e9-9604-de173b6bc044.png)